Pull request/ca0e90ba

jverzani opened the pull request:
Dear @isuruf, this pull request is similar in spirit to #18, but that got messed up. I'm closing that one and its sister request.

This keeps most objects as `Basic` objects, though does allow for dispatch on the SymEngine class through the `BasicType` class. The `diff` function and `N` function show a possible use case. Attempting to satisfy both Julia's type system and SymEngines at the same time proved difficult. I think this approach is a reasonable workaround.

There is one area that I could use some help with. The `ctypes.jl` file attempts to provide the CSetBasic and CVecBasic classes on the julia side so that `free_symbols` and `get_args` can be implemented. Both have issues: free symbols crashes right away, and get_args crashes eventually, though occasionally works as expected. I think I am not creating the finalizer correctly, but this is not an area I know much about.

isuruf commented on src/types.jl:
```diff
@@ -0,0 +1,200 @@
+
+## Class ID
+get_type(s::Basic) = convert(Int, ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s))
+get_class_from_id(id::Int) = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id) |> bytestring
```
`Ptr{UInt8}` will leak here right?

isuruf replied:
`basic_get_class_from_id` takes in a `UInt` not an `Int`

jverzani replied:
I don't really know, I don't program in C. I got this from
http://docs.julialang.org/en/release-0.4/manual/calling-c-and-fortran-code/
but if there is something else it should be, let me know.

On Fri, Apr 1, 2016 at 2:07 PM, Isuru Fernando <notifications@github.com>
wrote:

> In src/types.jl
> <https://github.com/symengine/SymEngine.jl/pull/20#discussion_r58244243>:
>
> > +
> > +if Clong == Int32
> > +    convert(::Type{Basic}, x::Union{Int8, Int16, Int32}) = Basic(convert(Clong, x))
> > +    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32}) = Basic(convert(Culong, x))
> > +else
> > +    convert(::Type{Basic}, x::Union{Int8, Int16, Int32, Int64}) = Basic(convert(Clong, x))
> > +    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32, UInt64}) = Basic(convert(Culong, x))
> > +end
> > +convert(::Type{Basic}, x::Integer) = Basic(BigInt(x))
> > +convert(::Type{Basic}, x::Rational) = Basic(num(x)) / Basic(den(x))
> > +
> > +Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
> > +
> > +## Class ID
> > +get_type(s::Basic) = convert(Int, ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s))
> > +get_class_from_id(id::Int) = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id) |> bytestring
>
> Ptr{UInt8} will leak here right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20/files/4b995dc4bba02cf4263c3a9daf197928ce187b16#r58244243>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf replied:
```
a = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id)
string = bytestring(a)
ccall((:basic_str_free, :libsymengine), Void, (Ptr{UInt8}, ), a)
string
```
should fix the leak

jverzani replied:
Thanks again!

On Sat, Apr 2, 2016 at 11:45 PM, Isuru Fernando <notifications@github.com>
wrote:

> In src/types.jl
> <https://github.com/symengine/SymEngine.jl/pull/20#discussion_r58304047>:
>
> > +
> > +if Clong == Int32
> > +    convert(::Type{Basic}, x::Union{Int8, Int16, Int32}) = Basic(convert(Clong, x))
> > +    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32}) = Basic(convert(Culong, x))
> > +else
> > +    convert(::Type{Basic}, x::Union{Int8, Int16, Int32, Int64}) = Basic(convert(Clong, x))
> > +    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32, UInt64}) = Basic(convert(Culong, x))
> > +end
> > +convert(::Type{Basic}, x::Integer) = Basic(BigInt(x))
> > +convert(::Type{Basic}, x::Rational) = Basic(num(x)) / Basic(den(x))
> > +
> > +Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
> > +
> > +## Class ID
> > +get_type(s::Basic) = convert(Int, ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s))
> > +get_class_from_id(id::Int) = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id) |> bytestring
>
> a = ccall((:basic_get_class_from_id, :libsymengine), Ptr{UInt8}, (Int,), id)
> string = bytestring(a)
> ccall((:basic_str_free, :libsymengine), Void, (Ptr{UInt8}, ), a)
> string
>
> should fix the leak
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20/files/4b995dc4bba02cf4263c3a9daf197928ce187b16#r58304047>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf commented on src/types.jl:
```diff
@@ -0,0 +1,200 @@
+Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
+
+## Class ID
+get_type(s::Basic) = convert(Int, ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s))
```
This should return a UInt

jverzani commented:
Thanks for the feedback! I adjusted two of the three things you suggested, and will do the other if there is something I should use otherwise. I also used your work on free_symbols to get lambdify to work. I added some exports, which seem reasonable to me.

jverzani commented:
This last commit added your fix for the memory leak. Thanks again! It also added a bit more to the interface, some of it hacky, but I wanted to be able to try some of the benchmarks from https://wiki.sagemath.org/symbench and certain things aren't yet wrapped in the cwrapper.cpp.

isuruf commented:
Are you getting the memory corruption errors when running tests as well?

jverzani commented:
No. Not from Pkg.test.

On Sunday, April 3, 2016, Isuru Fernando <notifications@github.com> wrote:

> Are you getting the memory corruption errors when running tests as well?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20#issuecomment-205021562>
>


-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf commented:
@jverzani, are you working on this branch or shall I merge this after travis passes?

jverzani commented:
No, I don't have any current additions. Thanks

On Mon, Apr 4, 2016 at 7:46 AM, Isuru Fernando <notifications@github.com>
wrote:

> @jverzani <https://github.com/jverzani>, are you working on this branch
> or shall I merge this after travis passes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20#issuecomment-205261115>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf commented:
@jverzani, thanks a lot for the PR.

jverzani commented:
Thank you for all the fixes! I'm especially sorry about the last bug
fix. You all are doing a great job here. Hopefully part of the GSOC will be
to wrap more functionality into libsymengine.

I'm really impressed by how much faster SymEngine is than SymPy running
under julia. I compared some of the examples here:
https://wiki.sagemath.org/symbench . When there are differences, they are
huge.


R1: 33.37microsecs; 25.77 ms  [SymEngine.jl, SymPy.jl]
R2: 27.24 ms;  6.26 s function may not have been precompiled
R3:  1.97 microsecs; 1.84 microsecs
R4: not implemented
R5: 11microsec; 3.04 ms
R6: no rational_simplify
R7: 171ms; 28s (precompile, as in complaint that function may not be
precompiled)
R8: 55ms; 13s (precompile)
R9: no `factor`
R10: 494 microsecs; 67.13 ms
R11: no sort on symbolic

S1: 30.51 ms;  310.60 μs;
S2: 35.91 ms; 13.56s (precompile),
S3:  26.74 ms; 32.55 ms
S4: 98.21 ms; 455.86 s (precompile)

On Mon, Apr 4, 2016 at 10:01 AM, Isuru Fernando <notifications@github.com>
wrote:

> @jverzani <https://github.com/jverzani>, thanks a lot for the PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20#issuecomment-205307058>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf commented:
Last bug was one introduced by me. :D.
I'm not that familiar with Julia, so the repo badly needed some love from a person with Julia experience.

[Here](https://github.com/symengine/symengine.py/blob/master/benchmarks/symbench.py)'s some of the symbench benchmarks in Python which benchmarks SymPy, Sage and SymEngine.py. 
See the [PR](https://github.com/symengine/symengine/pull/374) for timings

isuruf commented:
There seems to be something wrong in the timings for S1. Can you double check?

jverzani commented:
The output below is what I get from the following (using
https://github.com/johnmyleswhite/Benchmarks.jl)

```
module SymEngineTest

using SymEngine
x,y,z = symbols("x y z")


import Benchmarks

f = (x+y+z+1)^7;
expand(x);

Benchmarks.@benchmarkable(
    s1_benchmark,
    nothing,
    expand(f*(f+1)) ,
    nothing
)

r = Benchmarks.execute(s1_benchmark)
println("SymEngine")
println(r)


end

module SymPyTest


using SymPy
x,y,z = symbols("x y z")


import Benchmarks

f = (x+y+z+1)^7;
expand(x);

Benchmarks.@benchmarkable(
    s1_benchmark,
    nothing,
    expand(f*(f+1)) ,
    nothing
)

r = Benchmarks.execute(s1_benchmark)
println("SymPy")
println(r)

end
```


```

SymEngine

================ Benchmark Results ========================

     Time per evaluation: 30.50 ms [29.41 ms, 31.59 ms]

Proportion of time in GC: 0.00% [0.00%, 0.00%]

        Memory allocated: 16.00 bytes

   Number of allocations: 1 allocations

       Number of samples: 100

   Number of evaluations: 100

 Time spent benchmarking: 3.27 s



SymPy

================ Benchmark Results ========================

     Time per evaluation: 360.81 μs [334.90 μs, 386.72 μs]

Proportion of time in GC: 0.00% [0.00%, 0.00%]

        Memory allocated: 3.78 kb

   Number of allocations: 85 allocations

       Number of samples: 100

   Number of evaluations: 100

 Time spent benchmarking: 0.32 s

```


If it were helpful, I could wrap up the other benchmarks I could do like
this.


On Tue, Apr 5, 2016 at 9:38 AM, Isuru Fernando <notifications@github.com>
wrote:

> There seems to be something wrong in the timings for S1. Can you double
> check?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20#issuecomment-205808732>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf commented:
Benchmark should be like below
```
f = (x+y+z+1)^7;
f = expand(f);

Benchmarks.@benchmarkable(
    s1_benchmark,
    nothing,
    expand(f*(f+1)) ,
    nothing
)
```

jverzani commented:
Duh...  With this change I get:

```

julia> include("/tmp/test-s1.jl")

SymEngine

================ Benchmark Results ========================

     Time per evaluation: 30.60 ms [29.52 ms, 31.69 ms]

Proportion of time in GC: 0.00% [0.00%, 0.00%]

        Memory allocated: 16.00 bytes

   Number of allocations: 1 allocations

       Number of samples: 100

   Number of evaluations: 100

 Time spent benchmarking: 3.32 s


SymPy

================ Benchmark Results ========================

     Time per evaluation: 723.24 μs [694.99 μs, 751.49 μs]

Proportion of time in GC: 0.00% [0.00%, 0.00%]

        Memory allocated: 3.78 kb

   Number of allocations: 85 allocations

       Number of samples: 100

   Number of evaluations: 100

 Time spent benchmarking: 3.28 s

```



On Tue, Apr 5, 2016 at 10:13 AM, Isuru Fernando <notifications@github.com>
wrote:

> Benchmark should be like below
>
> f = (x+y+z+1)^7;
> f = expand(f);
>
> Benchmarks.@benchmarkable(
>     s1_benchmark,
>     nothing,
>     expand(f*(f+1)) ,
>     nothing
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20#issuecomment-205828335>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu

isuruf commented:
Hmm, maybe the SymPy call is optimized away and it is never called. In a Python shell, SymPy takes about 2s for the benchmark

jverzani commented:
Something like that for sure. It takes about 3 seconds the first time it is
called:

```


julia> using SymPy


julia> x,y,z = symbols("x y z")

(x,y,z)


julia> f2 = expand((x+y+z+1)^7);


julia> @time expand(f2 * (f2 + 1));

  2.909962 seconds (403 allocations: 52.511 KB)


julia> @time expand(f2 * (f2 + 1));

  0.014054 seconds (253 allocations: 10.578 KB)

```


SymEngine doesn't have that initial time:


```


julia> using SymEngine


julia> x,y,z = symbols("x y z")

(x,y,z)


julia> f2 = expand((x+y+z+1)^7);


julia> @time expand(f2 * (f2 + 1));

  0.035572 seconds (3.41 k allocations: 174.695 KB)

```

On Tue, Apr 5, 2016 at 10:32 AM, Isuru Fernando <notifications@github.com>
wrote:

> Hmm, maybe the SymPy call is optimized away and it is never called. In a
> Python shell, SymPy takes about 2s for the benchmark
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/20#issuecomment-205834945>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu
